### PR TITLE
Initial build infrastructure

### PR DESCRIPTION
**Pull request by justinsb**
Heavily based on k8s.io/registry.k8s.io

### CODE REVIEW

**Comment by ameukam** [comment]
@BenTheElder What the consensus on Github Actions vs Prow ?

/approve

File: dev/build-images [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+TAG="${TAG:-"$(date +v%Y%m%d)-$(git describe --always --dirty)"}"
+
+mkdir -p bin/
+
+# build images
+# TODO: bake commit info into binaries consistently
+for image in "${images[@]}"; do
+    name="$(basename "${image}")"
+    # push or local tarball
+    publish_args=(--tarball=bin/"${name}".tar --push=false)
```
Comment (ameukam) [inline dev/build-images]:
```suggestion
    publish_args=(--tarball=bin/"${name}".tar --platform=linux/amd64,linux/arm64 --push=false)
```

Don't cost a lot to publish arm64 images. Maybe useful in the future if we want to to run on thing like Lambda Graviton3.

Comment (justinsb) [reply dev/build-images]: Thanks - incorporated and marked you as a co-author!

**Comment by justinsb** [comment]
I had to tweak the `ko` command to not save tarballs, but instead to load locally into docker.  Otherwise we get an odd bug (?) in ko where it can't save the tarball with multiarch

```
2022/11/08 09:51:18 Using base distroless.dev/static:latest@sha256:4a58d5bcd0dcae2bfebf28f7ed9621c6d21838a5f343cdb9ee9450f1af478591 for sigs.k8s.io/porche/cmd/redirectserver
2022/11/08 09:51:18 Building sigs.k8s.io/porche/cmd/redirectserver for linux/arm64
2022/11/08 09:51:18 Building sigs.k8s.io/porche/cmd/redirectserver for linux/amd64
2022/11/08 09:51:19 Saving bin/redirectserver.tar
2022/11/08 09:51:19 failed to save "bin/redirectserver.tar": unable to calculate manifest: set of images is empty
Error: failed to publish images: error publishing ko://sigs.k8s.io/porche/cmd/redirectserver: failed to interpret sigs.k8s.io/porche/cmd/redirectserver result as image: &{0xc0001c4940 <nil> <nil> 0xc0001c2aa0 map[sbom:0xc0001c2a80]}
2022/11/08 09:51:19 error during command execution:failed to publish images: error publishing ko://sigs.k8s.io/porche/cmd/redirectserver: failed to interpret sigs.k8s.io/porche/cmd/redirectserver result as image: &{0xc0001c4940 <nil> <nil> 0xc0001c2aa0 map[sbom:0xc0001c2a80]}
exit status 1
```

**Comment by justinsb** [comment]
We can see the github actions here: https://github.com/justinsb/porche/pull/1

I figure we should use github actions (with minimal logic actually _in_ github actions) where we can, but for the image publishing or anything we need to be more trusted, those we run on prow.

**Comment by BenTheElder** [comment]
Actions don't provide as strong of guarantees for testing and I'm sure we can offload this particular CI to an EKS cluster in the future so ... shrug? Worth understanding the limitations of actions before committing:
- no way to /retest
- no way to guarantee latest changes in target and PR branch are tested before merge
- generally limited merge robot support

The CI for these repos is really small because we made the tests reasonable :-)

**Comment by justinsb** [comment]
OK, those seem like pretty good reasons to use prow :-)

I created a few scripts that are called by the prow jobs (proposed in https://github.com/kubernetes/test-infra/pull/27935)

Currently these are still called by github actions, but we can move to prow - just not sure of the sequencing here!

**Comment by ameukam** [comment]
/retest
(They must be some irony here)

/lgtm

**Comment by justinsb** [comment]
Had to push a small commit to tweak two things to make it work with prow:

    * Use sudo only if we're not root
    
    * Don't try to push to docker; instead only build for the local
      platform and export to a tarball.  ko doesn't seem to do well with
      local tarball + multiarch; we also don't have a local docker daemon
      in prow.

But prow is now happy :tada:

**Pull request merged**